components/brave_today/browser/direct_feed_controller.cc

```diff
@@ -86,33 +76,26 @@ mojom::ArticlePtr RustFeedItemToArticle(const FeedItem& rust_feed_item) {
 
 using ParseFeedCallback = base::OnceCallback<void(absl::optional<FeedData>)>;
 void ParseFeedDataOffMainThread(const GURL& feed_url,
-                                const std::string& charset,
                                 std::string body_content,
                                 ParseFeedCallback callback) {
   base::ThreadPool::PostTaskAndReplyWithResult(
       FROM_HERE,
       base::BindOnce(
-          [](const GURL& feed_url, const std::string& charset,
+          [](const GURL& feed_url,
              std::string body_content) -> absl::optional<FeedData> {
-            std::string result;
-            // Ensure the body is encoded as UTF-8.
-            if (!base::ConvertToUtf8AndNormalize(body_content, charset,
-                                                 &result)) {
-              LOG(ERROR) << "Failed to convert body content of " << feed_url
-                         << " to utf-8";
-              return absl::nullopt;
-            }
-
             brave_news::FeedData data;
-            if (!parse_feed_string(::rust::String(result), data)) {
+            if (!parse_feed_bytes(::rust::Slice<const uint8_t>(
+                                      (const uint8_t*)body_content.data(),
+                                      body_content.size()),
```

Comment:
I'm trying to wrap my head around, but quite not sure how slicing can make feed-rs stop decoding again :( could you tell me more about this?

Reply:
For sure, yeah! So we aren't stopping `feed_rs` from decoding at all, we're just passing the bytes array directly through to it, rather than forcing it into a string, and then [back into a bytes array](https://github.dev/brave/brave-core/blob/4af860a340c6f42b2c4753003972b136174a17a0/components/brave_today/rust/lib.rs#L34).

So before this PR, decode 1 was [here](https://github.dev/brave/brave-core/blob/4af860a340c6f42b2c4753003972b136174a17a0/components/brave_today/browser/direct_feed_controller.cc#L99) and decode two happened inside `feed_rs` [here](https://github.dev/brave/brave-core/blob/4af860a340c6f42b2c4753003972b136174a17a0/components/brave_today/rust/lib.rs#L34).

We've just removed decode 1 :smile:

Reply:
Aha! Thanks! I totally misunderstood this PR - I thought the opposite!